INT-3401: Make Google Guava as optional dep

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3401

Add Google Guava CLASSPATH check for the `StoredProcExecutor`
and use `ConcurrentHashMap` as an alternative to the Guava's `Cache`,
when Guava isn't presented in the CLASSPATH

artembilan committed:
INT-3401: Make Google Guava as optional dep

JIRA: https://jira.spring.io/browse/INT-3401

Add Google Guava CLASSPATH check for the `StoredProcExecutor`
and use `ConcurrentHashMap` as an alternative to the Guava's `Cache`,
when Guava isn't presented in the CLASSPATH

garyrussell commented on spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/StoredProcExecutor.java:
```diff
@@ -331,6 +353,20 @@ private String evaluateExpression(Message<?> message) {
+			SimpleJdbcCallOperations operations = this.jdbcCallOperationsMap.get(storedProcedureName);
+			if (operations == null) {
+				operations = createSimpleJdbcCall(storedProcedureName);
+				this.jdbcCallOperationsMap.putIfAbsent(storedProcedureName, operations);
```
We need a mechanism to limit the cache size. See #jdbcCallOperationsCacheSize.

Consider using a (synchronized) `LinkedHashMap` (ctor that takes the `accessOrder` boolean). See `LHM.removeEldestEntry()` - the map will take care of limiting the size.

garyrussell commented:
Needs a migration guide entry (tell users they have to manually add guava dep)

artembilan committed:
Addressing PR comments

artembilan commented:
Pushed.
Thank you, @garyrussell, for that `removeEldestEntry()` trick. It's exactly what I wanted to have to achieve the cache logic

garyrussell commented on spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/StoredProcExecutor.java:
```diff
@@ -153,6 +163,19 @@ public StoredProcExecutor(DataSource dataSource) {
 		this.dataSource = dataSource;
 
+		if (!guavaPresent) {
+			this.jdbcCallOperationsMap = new LinkedHashMap<String, SimpleJdbcCallOperations>() {
```
`SuppressWarnings("serial")`

Also, you need to use 

    new LinkedHashMap<String, SimpleJdbcCallOperations>(this.jdbcCallOperationsCacheSize + 1, 0.75f, true) {...}

to enable LRU processing.

garyrussell commented on spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/StoredProcExecutor.java:
```diff
@@ -210,23 +233,31 @@ public void afterPropertiesSet() {
+				this.jdbcCallOperationsCache.put(storedProcedureName, simpleJdbcCall);
+			}
+			else {
+				this.jdbcCallOperationsMap.put(storedProcedureName, simpleJdbcCall);
```
Should synchronize here too, to avoid Sonar barking at us (inconsistent sync).

That said, it's not clear to me why we preload this one.

artembilan committed:
PR Comments

artembilan commented:
Pushed

garyrussell commented:
Unfortunately this solution does not work.

There are two problems.

1. The anonymous subclass of `CacheLoader` causes it to be loaded, even though it's not needed.
     - fails in `getDeclaredConstructors()` while wiring up the SPE.
2. The `CacheStats` method return type causes it to be loaded in `getDeclaredMethods()`.

I tested this by wiring in a stored proc adapter into the jdbc sample, after removing the guava dep from there.

I think we're going to need a factory bean and return a subclass if guava is on the CP.

artembilan commented:
Doh... I had to do that testing before pushing...

OK. Let me come up with some solution!

artembilan committed:
PR Comments:

Introduce an internal static `StoredProcExecutor.GuavaCacheWrapper` class
to avoid the eager class loading for the `CacheLoader` anonymous class

artembilan commented:
Pushed.
Tested against samples

garyrussell commented:
Merged as 32e23d16420879f387481d174083786c959cdce1
